espnet2/enh/separator/tfgridnet_separator.py

```diff
@@ -0,0 +1,391 @@
+
+        batch = batch.view([n_batch, self.n_srcs, 2, n_frames, n_freqs])
+        batch = torch.complex(batch[:, :, 0], batch[:, :, 1])  # [B, n_srcs, T, F]
+
+        batch = self.dec(batch.view(-1, n_frames, n_freqs), ilens)[0]  # [B, n_srcs, -1]
+
+        batch = self.pad2(batch.view([n_batch, self.num_spk, -1]), n_samples)
+
+        batch = batch * mix_std_  # reverse the RMS normalization
+
+        batch = [batch[:, src] for src in range(self.num_spk)]
+
+        return batch, ilens, OrderedDict()
+
+    @property
+    def num_spk(self):
+        return self.n_srcs
+
+    @staticmethod
+    def pad2(input_tensor, target_len):
+        input_tensor = torch.nn.functional.pad(
+            input_tensor, (0, target_len - input_tensor.shape[-1])
+        )
+        return input_tensor
+
+
+class GridNetBlock(nn.Module):
+    def __getitem__(self, key):
+        return getattr(self, key)
+
+    def __init__(
+        self,
+        emb_dim,
+        emb_ks,
+        emb_hs,
+        n_freqs,
+        hidden_channels,
+        n_head=4,
+        approx_qk_dim=512,
+        activation="prelu",
+        eps=1e-5,
+    ):
+        super().__init__()
+
+        in_channels = emb_dim * emb_ks
+
+        self.intra_norm = LayerNormalization4D(emb_dim, eps=eps)
+        self.intra_rnn = nn.LSTM(
+            in_channels, hidden_channels, 1, batch_first=True, bidirectional=True
+        )
+        self.intra_linear = nn.ConvTranspose1d(
+            hidden_channels * 2, emb_dim, emb_ks, stride=emb_hs
+        )
+
+        self.inter_norm = LayerNormalization4D(emb_dim, eps=eps)
+        self.inter_rnn = nn.LSTM(
+            in_channels, hidden_channels, 1, batch_first=True, bidirectional=True
+        )
+        self.inter_linear = nn.ConvTranspose1d(
+            hidden_channels * 2, emb_dim, emb_ks, stride=emb_hs
+        )
+
+        E = math.ceil(
+            approx_qk_dim * 1.0 / n_freqs
+        )  # approx_qk_dim is only approximate
+        assert emb_dim % n_head == 0
+        for ii in range(n_head):
+            self.add_module(
+                "attn_conv_Q_%d" % ii,
+                nn.Sequential(
+                    nn.Conv2d(emb_dim, E, 1),
+                    get_layer(activation)(),
+                    LayerNormalization4DCF((E, n_freqs), eps=eps),
+                ),
+            )
+            self.add_module(
+                "attn_conv_K_%d" % ii,
+                nn.Sequential(
+                    nn.Conv2d(emb_dim, E, 1),
+                    get_layer(activation)(),
+                    LayerNormalization4DCF((E, n_freqs), eps=eps),
+                ),
+            )
+            self.add_module(
+                "attn_conv_V_%d" % ii,
+                nn.Sequential(
+                    nn.Conv2d(emb_dim, emb_dim // n_head, 1),
+                    get_layer(activation)(),
+                    LayerNormalization4DCF((emb_dim // n_head, n_freqs), eps=eps),
+                ),
+            )
```

Comment:
Hello, @zqwang7. I tried the code in this pull request, but it seems the code is slow. I guess these modules are the bottleneck. And merging these modules can largely increase the training speed.